_posts/2017-04-20-how-software-works-part1.md

```diff
@@ -0,0 +1,196 @@
+conform to a certain format that the CPU is wired to read and execute. For example, let's make up an instruction
+set with the following format:
+
+```
+1 byte | 1 byte | 1 byte
+-------+--------+-------
+OPCODE | NUMBER | NUMBER
+```
+
+The word opcode refers to what type of operation we want the CPU to execute on the given arguments. Now let's define
+what our imaginary architecture can do and what opcodes one would use to invoke those routines.
+
+```
+Opcode | Instruction name
+-------+-----------------
+0      | HALT
+1      | ADD
+2      | SUB
+3      | MUL
+4      | DIV
+```
+
+So, our computer can add, substract, multiply and divide two numbers. There's also a code for halting the execution
+for when the program has finished. A simple program for this machine would look like this:
+
+`00000001 00001010 00010100 00000000`
+
+To make this more human-readable, let's convert the binary digits back to decimal:
+
+`1 10 20 0`
+
+When we execute this machine code the computer reads in the first byte and sees that it needs to execute the
+`ADD` instruction, taking arguments `10` and `20`. Once that is done, it sees that it has to `HALT` the execution.
+The CPU halts, concluding our first program on our imaginary computer.
+
+I hope this section showed you how computers read _binary_ programs in a certain format and execute the
+specified instructions on _binary_ data. Now, in reality no-one really writes machine code by hand. It is
+just way too arduous to type out numbers with no mneumonics associated with them. This is why assembly was invented.
+
+### ADD 10 20 HALT
+
+Assembly, often abbreviated as _asm_, is the lowest-level code that humans usually write. It is a very small abstraction on top of machine code
+to make life a bit more convenient for us programmers. Here's how the same program would look like in assembly:
+
+```
+ADD 10 20
+HALT
+```
+
+This code will go through something called the _assembler_ which replaces the word `ADD` with the number 1
+and `HALT` with 0, producing the same binary file that we started with.
+
+Assembler is the first time we've encountered a _compiler_ on our journey to understand software. Compilers play a big
+role in being able to write code in higher-level languages. The job of a compiler is to transform a program from one
+language or format to another one. In this case, the assembler does that by replacing each occurrence of `ADD` or `HALT`
+with their corresponding binary version that the CPU can understand.
+
+Compilers aren’t limited to just doing converting programs, though. They often analyse the code and help the programmer avoid mistakes before transforming the code. For example, the assembler could analyse the code we've written and let us know if we've made a mistake:
+
+```
+ADD 10
+HALT
+```
+
+Seeing a program like that, the assembler might refuse to convert it into the binary equivalent because the code we've written does not conform to the specified opcode format.
```

Comment:
Seeing a program like that, the assembler might refuse to convert it into the binary equivalent because the code we've written does not conform to the specified opcode format (_the second argument needed in the addition is missing_).